modules/cache/cache.go

```diff
@@ -89,7 +89,16 @@ func (p *provider) starting(_ context.Context) error {
 }
 
 func (p *provider) stopping(_ error) error {
+	// we can only stop a cache once (or they panic). use this map
+	// to track which caches we've stopped.
+	stopped := map[cache.Cache]struct{}{}
```

Comment:
Yup. If a cache has multiple roles we add it to the map multiple times here:

https://github.com/grafana/tempo/blob/8d0ddfd2d35d460b56ce44cc2b09690c0abdff03/modules/cache/cache.go#L60-L62

So this is just a bit of bookkeeping to prevent stopping one twice.